File: abc/python/abc215/abc215_H.py

import sys
#import random
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

MOD = 998244353

## Adapted from https://codeforces.com/blog/entry/45223
## Given an array A of length (1<<N) that represent a function of a subset of a set of length (N)
## Return a new array F that also represents a function of a subset of a set of length(N)
##    F(bitmask) = sum_over_bm2=subsets_of_bitmask A(bm2)
def sumoversubsets(N,A,inplace=False) :
    F = A if inplace else A.copy()
    for i in range(N) :
        for mask in range(1<<N) :
            if mask & (1<<i) :
                F[mask] += F[mask^(1<<i)]
    if not inplace : return F

def sumoversupersets(N,A,inplace=False) :
    if inplace : 
        A.reverse()
        sumoversubsets(N,A,True)
        A.reverse()
    else :
        F = A.copy()
        F.reverse()
        sumoversubsets(N,F,True)
        F.reverse()
        return F
    
## AKA Mobius Transform
## https://codeforces.com/blog/entry/72488
def inversesumoversubsets(N,A,inplace=False) :
    F = A if inplace else A.copy()
    for i in range(N) :
        for mask in range(1<<N) :
            if mask & (1<<i) :
                F[mask] -= F[mask^(1<<i)]
                F[mask] %= MOD
    if not inplace : return F

def solve(N,M,A,B,C) :
    ## First we need to figure out how many cabbages Snuke needs to eat
    ## Use Hall's Marriage Theorem and SumOfSubsets to figure this out.

    ## Calculate the supply for each bitmask
    supply = [0] * (1<<N)
    for bm in range(1<<N) :
        for n in range(N) :
            if bm & (1<<n) != 0 : supply[bm] += A[n]

    ## Calculate the demand for each order individually, and then do sum over subsets to aggregate orders
    demand = [0] * (1<<N)
    for m in range(M) :
        bm = 0
        for n in range(N) :
            if C[n][m] : bm |= (1<<n)
        demand[bm] += B[m]
    sumoversubsets(N,demand,True) 

    ## Now find the smallest excess
    myinf = 10**18
    minexcess = min(myinf if demand[x] == 0 else supply[x]-demand[x] for x in range(1<<N))

    ## If excess is less than zero, then Snuke doesn't need to eat anything.  Otherwise, he needs to eat 1 more than the excess
    if minexcess < 0 : return "0 1"
    snuke = minexcess+1

    ## Find the bitmasks which exercise this minexcess, and then propagate this to subsets
    feasible = [1 if demand[x] > 0 and supply[x]-demand[x] == minexcess else 0 for x in range(1<<N)]
    sumoversupersets(N,feasible,True)
    
    ## Calculate the ways to satisfy snuke within each subset
    maxsupply = max(supply)
    fact = [1] * (maxsupply+1)
    factinv = [1] * (maxsupply+1)
    binom = [0] * (maxsupply+1)
    for i in range(2,maxsupply+1) : fact[i] = i * fact[i-1] % MOD
    factinv[maxsupply] = pow(fact[maxsupply],MOD-2,MOD)
    for i in range(maxsupply-1,-1,-1) : factinv[i] = (i+1) * factinv[i+1] % MOD
    for x in range(snuke,maxsupply+1) : binom[x] = fact[x] * factinv[snuke] % MOD * factinv[x-snuke] % MOD
    uniqueways = [ 0 if feasible[bm] == 0 else binom[supply[bm]] for bm in range (1<<N)]
    inversesumoversubsets(N,uniqueways,True)
    ## Apply Mobius (aka inverse SoS) to find the unique ways for each subset that aren't included in child subsets
    snukeways = sum(uniqueways[x] * (1 if feasible[x] else 0) for x in range(1<<N)) % MOD
    return f"{snuke} {snukeways}"

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    N,M = gis()
    A = gis()
    B = gis()
    C = []
    for _ in range(N) : C.append(gis())
    ans = solve(N,M,A,B,C)
    sys.stdout.write(str(ans)+'\n')

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

File: abc/go/abc215/abc215_H/abc215_H.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)
var wrtr = bufio.NewWriterSize(os.Stdout, 10_000_000)
var rdr = bufio.NewScanner(os.Stdin)
func gs() string  { rdr.Scan(); return rdr.Text() }
func gi() int     { i,e := strconv.Atoi(gs()); if e != nil {panic(e)}; return i }
func gi2() (int,int) { return gi(),gi() }
func gis(n int) []int  { res := make([]int,n); for i:=0;i<n;i++ { res[i] = gi() }; return res }
func ia(m int) []int { return make([]int,m) }
func rev(a []int) { i,j := 0,len(a)-1; for i < j { a[i],a[j] = a[j],a[i]; i++; j-- } }
func min(a,b int) int { if a > b { return b }; return a }
func maxarr(a []int) int { ans := a[0]; for _,aa := range(a) { if aa > ans { ans = aa } }; return ans }
func powmod(a,e,mod int) int { res, m := 1, a; for e > 0 { if e&1 != 0 { res = res * m % mod }; m = m * m % mod; e >>= 1 }; return res }
func makefact(n int,mod int) ([]int,[]int) {
	fact,factinv := make([]int,n+1),make([]int,n+1)
	fact[0] = 1; for i:=1;i<=n;i++ { fact[i] = fact[i-1] * i % mod }
	factinv[n] = powmod(fact[n],mod-2,mod); for i:=n-1;i>=0;i-- { factinv[i] = factinv[i+1] * (i+1) % mod }
	return fact,factinv
}

const MOD int = 998244353

// Adapted from https://codeforces.com/blog/entry/45233
func sumoversubsets(N int, A []int, inplace bool) []int {
	F := A; if !inplace { F = make([]int,len(A)); copy(F,A) }
	for i:=0;i<N;i++ {
		for mask:=0;mask<1<<N;mask++ {
			if mask & (1<<i) != 0 { F[mask] += F[mask^(1<<i)] }
		}
	}
	return F
}

func sumoversupersets(N int, A []int, inplace bool) []int {
	F := A; if !inplace { F = make([]int,len(A)); copy(F,A) }
	rev(F); sumoversubsets(N,F,true); rev(F); return F
}

// AKA Mobius Transorm https://codeforces.com/blog/entry/72488
func inversesumoversubsets(N int, A []int, inplace bool) []int {
	F := A; if !inplace { F = make([]int,len(A)); copy(F,A) }
	for i:=0;i<N;i++ {
		for mask:=0;mask<1<<N;mask++ {
			if mask & (1<<i) != 0 { 
				F[mask] += MOD - F[mask^(1<<i)]
				F[mask] %= MOD
			}
		}
	}
	return F
}

func main() {
	//f1, _ := os.Create("cpu.prof"); pprof.StartCPUProfile(f1); defer pprof.StopCPUProfile()
	defer wrtr.Flush()
	infn := ""; if infn == "" && len(os.Args) > 1 {	infn = os.Args[1] }
	if infn != "" {	f, e := os.Open(infn); if e != nil { panic(e) }; rdr = bufio.NewScanner(f) }
	rdr.Split(bufio.ScanWords); rdr.Buffer(make([]byte,1024),1_000_000_000)
	// PROGRAM STARTS HERE
	N,M := gi2(); A := gis(N); B := gis(M)
	C := make([][]int,N); for i:=0;i<N;i++ { C[i] = gis(M) }
	supply := ia(1<<N)
	for bm:=0;bm<1<<N;bm++ {
		for n:=0;n<N;n++ {
			if bm & (1<<n) != 0 { supply[bm] += A[n] }
		}
	}
	demand := ia(1<<N)
	for m:=0;m<M;m++ {
		bm := 0
		for n:=0;n<N;n++ { if C[n][m] != 0 { bm |= (1<<n) } }
		demand[bm] += B[m]
	}
	sumoversubsets(N,demand,true)
	inf := 1_000_000_000_000_000_000
	minexcess := inf
	for x:=0;x<1<<N;x++ { if demand[x] > 0 { minexcess = min(minexcess,supply[x]-demand[x])}}
	if minexcess < 0 { fmt.Println("0 1"); return }
	snuke := minexcess+1
	feasible := ia(1<<N)
	for x:=0;x<1<<N;x++ { if demand[x] > 0 && supply[x]-demand[x] == minexcess { feasible[x] = 1 } }
	sumoversupersets(N,feasible,true)
	maxsupply := maxarr(supply)
	fact,factinv := makefact(maxsupply,MOD)
	binom := ia(maxsupply+1)
	for x:=snuke;x<=maxsupply;x++ { binom[x] = fact[x] * factinv[snuke] % MOD * factinv[x-snuke] % MOD }
	uniqueways := ia(1<<N)
	for x:=0;x<1<<N;x++ { if feasible[x] > 0 { uniqueways[x] = binom[supply[x]] } }
	inversesumoversubsets(N,uniqueways,true)
	snukeways := 0
	for x:=0;x<1<<N;x++ { if feasible[x] > 0 { snukeways += uniqueways[x] } }
	snukeways %= MOD
	fmt.Printf("%v %v\n",snuke,snukeways)
}

